test/unit/file.test.ts

```diff
@@ -0,0 +1,103 @@
+/*
+ * sonar-scanner-npm
+ * Copyright (C) 2022-2024 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import fs from 'fs';
+import * as fsExtra from 'fs-extra';
+import path from 'path';
+import AdmZip from 'adm-zip';
+import zlib from 'zlib';
+import tarStream from 'tar-stream';
+import { extractArchive, getCachedFileLocation } from '../../src/file';
+import { log, LogLevel } from '../../src/logging';
+import { SONAR_CACHE_DIR } from '../../src/constants';
+import { Readable } from 'stream';
+
+// Mock the filesystem
+jest.mock('fs', () => ({
+  createReadStream: jest.fn().mockImplementation(() => {
+    const mockStream = new Readable({
+      read() {
+        process.nextTick(() => this.emit('end')); // emit 'end' on next tick
+      },
+    });
+    mockStream.pipe = jest.fn().mockReturnThis();
+    return mockStream;
+  }),
+  createWriteStream: jest.fn(),
+  existsSync: jest.fn(),
+}));
+
+jest.mock('fs-extra', () => ({}));
+
+jest.mock('adm-zip', () => {
+  const MockAdmZip = jest.fn();
+  MockAdmZip.prototype.extractAllTo = jest.fn();
+  return MockAdmZip;
+});
+
+jest.mock('tar-stream', () => ({
+  extract: jest.fn().mockReturnValue({
+    on: jest.fn(),
+  }),
+}));
```

Comment:
this isn't used atm but could be:
```suggestion
jest.mock('tar-stream', () => ({
  extract: jest.fn().mockReturnValue({
    on: jest.fn().mockImplementation((event, cb) => {
      if (event === 'finish') {
        cb();
      }
    }),
  }),
}));
```